inverted_index.py

```diff
@@ -0,0 +1,354 @@
+                            self.postings[term].append(str(doc_id))
+
+                        curr_block_size += 1
+
+                        # Write the previous items to new block
+                        if curr_block_size == self.BLOCK_SIZE_LIMIT:
+                            block_index += 1
+                            self.WriteBlockToDisk(block_index)
+                            curr_block_size = 0
+
+
+        self.MergeBlocks(block_index)
+
+
+    """
+               Method to create /blocks folder to keep all the intermediate block files
+               Method to remove all existing /block folders and posting.txt and dictionary.txt files   
+    """
+    def ResetFiles(self):
+
+        block_dir = "blocks"
+        if os.path.exists(block_dir):
+            # Delete block folder
+            shutil.rmtree(block_dir)
+        os.makedirs(block_dir)
+
+        if os.path.exists(self.out_postings):
+            os.remove(self.out_postings)
+
+        if os.path.exists(self.out_dict):
+            os.remove(self.out_dict)
+
+    """
+            Method to write the contents of posting lists (stored as self.postings in memory) to a new block file
+            SPIMI-Invert
+            Params:
+                - block_index: Gives us the file name
+    """
+    def WriteBlockToDisk(self, block_index):
+        print("create new block ... " + str(block_index))
+
+        result = ""
+        # Sort Dictionary Terms. Line 11 of SPIMI - INVERT
+        for term in sorted(self.dictionary):
+            result += term + " "
+            # No need to sort postings because they are already in sorted form when we processed
+            # the docs in ascending order previously
+            for doc_id in self.postings[term]:
+                result += doc_id + " "
+            result += "\n"
+
+        self.WriteToFile("blocks/"+str(block_index), result)
+
+        # Clear Postings and Dictionary from memory
+        self.postings = defaultdict(list)
+        self.dictionary = set()
+
+    """
+            Method to read all the block files inside /Blocks and append them into Posting.txt
+            using BSBI Merging -> (total_num_blocks) K-Way Merge
+            Params:
+                - total_num_blocks: Total number of block files that we want to merge
+    """
+    def MergeBlocks(self, total_num_blocks):
+        print("Merge all blocks ...")
+
+        # blocks_offset[i] stores the offset value for reading from file of block i
+        blocks_offset = [0 for x in range(0, total_num_blocks+1)] # ignore value at index 0
+
+        q = PriorityQueue()
+
+        # Add the first line of each block to priority queue
+        for block_num in range(1, total_num_blocks + 1):
+            line = self.ReadFromFile("blocks/"+str(block_num), blocks_offset[block_num])
+            blocks_offset[block_num] = blocks_offset[block_num] + len(line) + 1 # 1 for \n
+            q.put(QueueItem(line, block_num))
+
+        # K-Way
```

Comment:
implicit assumption here that we can hold K pages in memory - might need to implement multi-pass K way merge

Reply:
TODO